onnx/defs/nn/defs.cc

```diff
@@ -1534,37 +1534,49 @@ ONNX_OPERATOR_SET_SCHEMA(
             "var",
             "running (training) or estimated (testing) variance tensor of shape (C).",
             "T")
+        .Input(
+            5,
+            "training_mode",
+            "If set to true, run spatial batch normalization in training mode, default is false.",
+            "T1",
+            OpSchema::Optional)
```

Comment:
If we make it an attribute, we can do static shape inference for both of training and inference. If we make it an input, I am not sure how ONNX shape inference would work.

Reply:
If we make it an attribute, then training_mode would be fixed once the model is exported.
So if we export the model in training_mode, we would not be able to run the model in inference mode since batch_norm will have a fixed attribute value of training_mode.
I am thinking that this should not be an issue, is that correct? In that case, I will change it to be an attribute instead.

Reply:
My original concern is  its dynamic output schema which depends on an input flag. Can we make them optional (like LSTM's outputs)? It can probably make ONNX's static shape inference working again.

Reply:
This needs to be an an input following the comment: https://github.com/onnx/onnx/pull/2568#discussion_r371459340 .
You mean make the output mean/var optional? It is already the case.

Reply:
But you mentioned that if `training_mode=True`, those outputs won't be populated. This doesn't match the behavior of LSTM --- A optional output will be always populated as long as the user specifies that in the output list.

Reply:
got it. I updated the doc for them to be optional

Reply:
I didn't understand the last two comments. For output values that will not be used in training_mode=false, the spec should not force the computation of those values.

Reply:
More specifically: when training_mode=false, which output-values need to be computed?

Reply:
'mean', 'var', 'saved_mean', and 'saved_var' are only needed for training.
But they are optional parameters (so they have static shapes and we can do the shape inference), so even if training_mode=False and the user specify them as output, they will be outputs of the op